test/decompiler/reference/levels/common/nav-enemy_REF.gc

```diff
@@ -0,0 +1,3973 @@
+          (-> self collide-info trans)
+          (-> *target* control trans)
+          )
+         )
+        )
+       )
+      (logior! (-> self nav-enemy-flags) 1)
+      (set! (-> self notice-time) (-> *display* base-frame-counter))
+      )
+     )
+    )
+   gp-0
+   )
+  )
+
+;; definition for function nav-enemy-facing-direction?
+;; Used lq/sq
+(defbehavior nav-enemy-facing-direction? nav-enemy ((arg0 vector) (arg1 float))
+  (let
+   ((s4-0
+     (vector-z-quaternion!
+      (new 'stack-no-clear 'vector)
+      (-> self collide-info quat)
+      )
+     )
+    (s5-0 (new 'stack-no-clear 'vector))
+    )
+   (set! (-> s5-0 quad) (-> arg0 quad))
+   (set! (-> s5-0 y) 0.0)
+   (vector-normalize! s5-0 1.0)
+   (>= (vector-dot s4-0 s5-0) (cos arg1))
+   )
+  )
+
+;; definition for function nav-enemy-facing-point?
+(defbehavior nav-enemy-facing-point? nav-enemy ((arg0 vector) (arg1 float))
+  (let
+   ((v1-1
+     (vector-! (new 'stack-no-clear 'vector) arg0 (-> self collide-info trans))
+     )
+    )
+   (nav-enemy-facing-direction? v1-1 arg1)
+   )
+  )
+
+;; definition for function nav-enemy-facing-player?
+(defun nav-enemy-facing-player? ((arg0 float))
+  (nav-enemy-facing-point? (target-pos 0) arg0)
+  )
+
+;; definition for function nav-enemy-test-nav-mesh-intersection-xz?
+(defbehavior nav-enemy-test-nav-mesh-intersection-xz? nav-enemy ((arg0 vector))
+  (dummy-16 (-> self nav) arg0)
+  )
+
+;; definition for function nav-enemy-test-point-in-nav-mesh?
+(defbehavior nav-enemy-test-point-in-nav-mesh? nav-enemy ((arg0 vector))
+  (and
+   (dummy-16 (-> self nav) arg0)
+   (<
+    (-> arg0 y)
+    (+ (-> self collide-info trans y) (-> self enemy-info notice-top))
+    )
+   )
+  )
+
+;; definition for function nav-enemy-test-point-near-nav-mesh?
+(defbehavior nav-enemy-test-point-near-nav-mesh? nav-enemy ((arg0 vector))
+  (and
+   (dummy-25 (-> self nav) arg0 (-> self nav-info notice-nav-radius))
+   (<
+    (-> arg0 y)
+    (+ (-> self collide-info trans y) (-> self enemy-info notice-top))
+    )
+   )
+  )
+
+;; definition for function nav-enemy-get-new-patrol-point
+(defbehavior nav-enemy-get-new-patrol-point nav-enemy ()
+  (TODO-RENAME-42 self)
+  0
+  )
+
+;; definition for function nav-enemy-set-base-collide-sphere-collide-with
+;; INFO: Return type mismatch int vs none.
+(defbehavior
+  nav-enemy-set-base-collide-sphere-collide-with nav-enemy
+  ((arg0 uint))
+  (let ((s5-0 (-> self collide-info root-prim)))
+   (set! (-> s5-0 collide-with) arg0)
+   (let
+    ((s4-0
+      (if
+       (and
+        (nonzero? s5-0)
+        (type-type? (-> s5-0 type) collide-shape-prim-group)
+        )
+       s5-0
+       )
+      )
+     (s5-1 (the-as object #f))
+     )
+    (when s4-0
+     (dotimes (s3-0 (-> (the-as collide-shape-prim-group s4-0) num-prims))
+      (let*
+       ((s2-0
+         (the-as
+          object
+          (->
+           (the-as collide-shape-prim-group (+ (* s3-0 4) (the-as int s4-0)))
```

Comment:
this looks like a bad array access, we might need to fix the `collide-shape-prim-group` type

Reply:
Yeah this method is a tricky one as the `collide-shape-prim` can be multiple things at runtime.  im probably missing some casts (or have too many)

Reply:
okay, let's leave it like this for now then, until we do the collision stuff.